File: modules/markup/external/viewer3d.go
// Copyright 2026 The Gitea Authors. All rights reserved.
// SPDX-License-Identifier: MIT

package external

import (
	"encoding/base64"
	"fmt"
	"html"
	"io"
	"path"

	"code.gitea.io/gitea/modules/markup"
	"code.gitea.io/gitea/modules/public"
	"code.gitea.io/gitea/modules/setting"
)

type viewer3dRenderer struct{}

var (
	_ markup.PostProcessRenderer = (*viewer3dRenderer)(nil)
	_ markup.ExternalRenderer    = (*viewer3dRenderer)(nil)
)

func (p *viewer3dRenderer) Name() string {
	return "viewer3d"
}

func (p *viewer3dRenderer) NeedPostProcess() bool {
	return false
}

func (p *viewer3dRenderer) FileNamePatterns() []string {
	return []string{
		"*.3dm", "*.3ds", "*.3mf", "*.amf", "*.bim", "*.brep",
		"*.dae", "*.fbx", "*.fcstd", "*.glb", "*.gltf",
		"*.ifc", "*.igs", "*.iges", "*.stp", "*.step",
		"*.stl", "*.obj", "*.off", "*.ply", "*.wrl",
	}
}

func (p *viewer3dRenderer) SanitizerRules() []setting.MarkupSanitizerRule {
	return nil
}

func (p *viewer3dRenderer) GetExternalRendererOptions() (ret markup.ExternalRendererOptions) {
	ret.SanitizerDisabled = true
	ret.DisplayInIframe = true
	ret.ContentSandbox = "allow-scripts"
	return ret
}

func (p *viewer3dRenderer) Render(ctx *markup.RenderContext, input io.Reader, output io.Writer) error {
	if ctx.RenderOptions.StandalonePageOptions == nil {
		opts := p.GetExternalRendererOptions()
		return markup.RenderIFrame(ctx, &opts, output)
	}

	if _, err := fmt.Fprintf(output,
		`<!DOCTYPE html>
<html>
<head><meta charset="utf-8"><style>html,body{margin:0;height:100%%}#viewer{width:100%%;height:100%%}#viewer>div{text-align:center;padding:1em}</style></head>
<body><div id="viewer"></div><script type="application/octet-stream" id="modelData" data-filename="%s">`,
		html.EscapeString(path.Base(ctx.RenderOptions.RelativePath)),
	); err != nil {
		return err
	}

	b64 := base64.NewEncoder(base64.StdEncoding, output)
	if _, err := io.Copy(b64, io.LimitReader(input, setting.UI.MaxDisplayFileSize)); err != nil {
		return err
	}
	if err := b64.Close(); err != nil {
		return err
	}

	_, err := fmt.Fprintf(output, `</script><script type="module" src="%s"></script></body></html>`, public.AssetURI("js/viewer-3d.js"))
	return err
}

File: web_src/js/render/plugins/viewer3d.ts
import type {FileRenderPlugin} from '../plugin.ts';
import {basename, extname} from '../../utils.ts';
import {getRealBackgroundColor} from '../../markup/render-iframe.ts';
import {GET} from '../../modules/fetch.ts';
import {html} from '../../utils/html.ts';

const SUPPORTED_EXTENSIONS = [
  '.3dm', '.3ds', '.3mf', '.amf', '.bim', '.brep',
  '.dae', '.fbx', '.fcstd', '.glb', '.gltf',
  '.ifc', '.igs', '.iges', '.stp', '.step',
  '.stl', '.obj', '.off', '.ply', '.wrl',
];

export function newRenderPlugin3DViewer(): FileRenderPlugin {
  return {
    name: '3d-model-viewer',

    canHandle(filename: string, _mimeType: string): boolean {
      return SUPPORTED_EXTENSIONS.includes(extname(filename).toLowerCase());
    },

    // Render inside a sandboxed iframe (`allow-scripts` without `allow-same-origin` → null origin)
    // so the 3D library runs isolated from the parent. The parent fetches the file because the
    // iframe can't read same-origin URLs, and ships bytes+bgcolor over postMessage.
    async render(container: HTMLElement, fileUrl: string): Promise<void> {
      const viewerUrl = container.getAttribute('data-viewer3d-url')!;
      const bgcolor = getRealBackgroundColor(container);
      const primary = getComputedStyle(document.documentElement).getPropertyValue('--color-primary').trim();

      const iframe = document.createElement('iframe');
      iframe.setAttribute('sandbox', 'allow-scripts');
      iframe.className = 'tw-w-full tw-h-full tw-border-0';
      iframe.srcdoc = html`
        <!DOCTYPE html>
        <html><head><meta charset="utf-8"><style>html,body{margin:0;height:100%;overflow:hidden;background:${bgcolor}}#viewer{width:100%;height:100%}</style></head><body><div id="viewer"></div><script type="module" src="${viewerUrl}"></script></body></html>
      `;
      container.append(iframe);

      const [response] = await Promise.all([
        GET(fileUrl),
        new Promise((resolve) => iframe.addEventListener('load', resolve, {once: true})),
      ]);
      if (!response.ok) throw new Error(`failed to fetch file: ${response.status}`);
      const bytes = await response.arrayBuffer();
      iframe.contentWindow!.postMessage({filename: basename(fileUrl), bytes, bgcolor, primary}, '*', [bytes]);
    },
  };
}
